src/cc/bcc_proc.c

```diff
@@ -392,3 +392,74 @@ char *bcc_procutils_which_so(const char *libname, int pid) {
 void bcc_procutils_free(const char *ptr) {
   free((void *)ptr);
 }
+
+/* Detects the following languages + C. */
+const char *languages[] = {"java", "python", "ruby", "php", "node"};
+const int nb_languages = 5;
+
+char *bcc_procutils_language(int pid) {
+  char procfilename[22], line[4096], pathname[32], *str;
+  FILE *procfile;
+  int i, ret;
+
+  /* Look for clues in the process' command line. */
+  sprintf(procfilename, "/proc/%ld/cmdline", (long)pid);
+  procfile = fopen(procfilename, "r");
+  if (!procfile)
+    return NULL;
+
+  ret = fscanf(procfile, "%s", line);
+  for (i = 0; i < nb_languages; i++)
+    if (strstr(line, languages[i]))
+      return strdup(languages[i]);
+
+  fclose(procfile);
+
+  /* Look for clues in the absolute path to the executable. */
+  sprintf(procfilename, "/proc/%ld/exe", (long)pid);
+  if (realpath(procfilename, line)) {
+    for (i = 0; i < nb_languages; i++)
+      if (strstr(line, languages[i]))
+        return strdup(languages[i]);
+  }
+
+
+  sprintf(procfilename, "/proc/%ld/maps", (long)pid);
+  procfile = fopen(procfilename, "r");
+  if (!procfile)
+    return NULL;
+
+  /* Look for clues in memory mappings. */
+  bool libc = false;
+  do {
+    char perm[8], dev[8];
+    long long begin, end, size, inode;
+    ret = fscanf(procfile, "%llx-%llx %s %llx %s %lld", &begin, &end, perm,
+                 &size, dev, &inode);
+    if (!fgets(line, sizeof(line), procfile))
+      break;
+    if (ret == 6) {
+      char *mapname = line;
+      char *newline = strchr(line, '\n');
+      if (newline)
+        newline[0] = '\0';
+      while (isspace(mapname[0])) mapname++;
+      for (i = 0; i < nb_languages; i++) {
+        sprintf(pathname, "/bin/%s", languages[i]);
+        if (strstr(mapname, pathname))
+          return strdup(languages[i]);
```

Comment:
Is this (looking for `/bin/$language`) the same as checking the absolute path of `/proc/$pid/exe`?